datafusion/core/src/sql/planner.rs

```diff
@@ -296,9 +296,10 @@ impl<'a, S: ContextProvider> SqlToRel<'a, S> {
 
         let plan = self.order_by(plan, query.order_by)?;
 
-        let plan: LogicalPlan = self.offset(plan, query.offset)?;
+        let plan: LogicalPlan = self.limit(plan, query.limit)?;
 
-        self.limit(plan, query.limit)
+        //make limit as offset's input will enable limit push down simply
+        self.offset(plan, query.offset)
```

Comment:
If there is an `offset` in sql, we will put it as root node in plan tree. So it will easily push `new_limit` to limit operator.

Reply:
Don't we need to apply the offset before the limit here?

It looks like we don't currently have a test with both a non-zero offset and a limit. Could you add one here so we can see what the plan looks like?

Per https://www.postgresql.org/docs/current/queries-limit.html `If both OFFSET and LIMIT appear, then OFFSET rows are skipped before starting to count the LIMIT rows that are returned.` and I am not sure this change will ensure that

Reply:
form the explain from pg
```
postgres=# explain analyze  select * from users limit 5 offset 7;
                                                 QUERY PLAN
-------------------------------------------------------------------------------------------------------------
 Limit  (cost=0.17..0.29 rows=5 width=80) (actual time=0.117..0.226 rows=5 loops=1)
   ->  Seq Scan on users  (cost=0.00..239.02 rows=10002 width=80) (actual time=0.013..0.082 rows=12 loops=1)
 Planning Time: 0.060 ms
 Execution Time: 0.350 ms
(4 rows)
```
I think in PG, the limit operator has a param with `offset`， in DF we separate it. So we need apply offset after limit `a+b`

Reply:
fix and add test in [20366bd](https://github.com/apache/arrow-datafusion/pull/2566/commits/20366bd32654ee0691bae3598f4f41be7960de48)